packages/fuels-abigen-macro/src/setup_contract_test_macro/parsing.rs

```diff
@@ -0,0 +1,228 @@
+use std::collections::HashSet;
+
+use itertools::{chain, Itertools};
+use proc_macro2::Span;
+use syn::{
+    parse::{Parse, ParseStream},
+    Error, LitStr, Result as ParseResult,
+};
+
+use crate::parse_utils::{Command, ErrorsExt, UniqueLitStrs, UniqueNameValues};
+
+trait MacroName {
+    fn macro_command_name() -> &'static str;
+}
+
+pub(crate) struct InitializeWallet {
+    pub(crate) span: Span,
+    pub(crate) names: Vec<LitStr>,
+}
+
+impl MacroName for InitializeWallet {
+    fn macro_command_name() -> &'static str {
+        "Wallets"
+    }
+}
+
+impl TryFrom<Command> for InitializeWallet {
+    type Error = Error;
+
+    fn try_from(command: Command) -> Result<Self, Self::Error> {
+        validate_command_has_correct_name::<Self>(&command)?;
+
+        let wallets = UniqueLitStrs::new(command.contents)?;
+
+        Ok(Self {
+            span: command.name.span(),
+            names: wallets.into_iter().collect(),
+        })
+    }
+}
+
+pub(crate) struct GenerateContract {
+    pub(crate) name: LitStr,
+    pub(crate) abi: LitStr,
+}
+
+impl MacroName for GenerateContract {
+    fn macro_command_name() -> &'static str {
+        "Abigen"
+    }
+}
+
+impl TryFrom<Command> for GenerateContract {
+    type Error = Error;
+
+    fn try_from(command: Command) -> Result<Self, Self::Error> {
+        validate_command_has_correct_name::<Self>(&command)?;
+
+        let name_values = UniqueNameValues::new(command.contents)?;
+        name_values.validate_has_no_other_names(&["name", "abi"])?;
+
+        let name = name_values.get_as_lit_str("name")?.clone();
+        let abi = name_values.get_as_lit_str("abi")?.clone();
+
+        Ok(Self { name, abi })
+    }
+}
+
+pub(crate) struct DeployContract {
+    pub name: String,
+    pub contract: LitStr,
+    pub wallet: String,
+}
+
+impl MacroName for DeployContract {
+    fn macro_command_name() -> &'static str {
+        "Deploy"
+    }
+}
+
+impl TryFrom<Command> for DeployContract {
+    type Error = Error;
+
+    fn try_from(command: Command) -> Result<Self, Self::Error> {
+        validate_command_has_correct_name::<Self>(&command)?;
+        let name_values = UniqueNameValues::new(command.contents)?;
+        name_values.validate_has_no_other_names(&["name", "contract", "wallet"])?;
+
+        let name = name_values.get_as_lit_str("name")?.value();
+        let contract = name_values.get_as_lit_str("contract")?.clone();
+        let wallet = name_values.get_as_lit_str("wallet")?.value();
+
+        Ok(Self {
+            name,
+            contract,
+            wallet,
+        })
+    }
+}
+
+fn validate_command_has_correct_name<T: MacroName>(command: &Command) -> syn::Result<()> {
+    let expected_name = T::macro_command_name();
+    if command.name == expected_name {
+        Ok(())
+    } else {
+        Err(Error::new_spanned(
+            command.name.clone(),
+            format!("Expected command to have name: '{expected_name}'."),
+        ))
+    }
+}
```

Comment:
Great idea, doing it now.

Reply:
I've pushed the closest approach that comes to mind. A bit cleaner, if nothing else.

It would be a bit weird if Command had this validation since it doesn't care about it, any name is a valid name as far as Command is concerned.

Also, it would introduce a cycle in the dependencies, would need more files to break apart.